src/phase0/helpers.rs

```diff
@@ -97,48 +97,40 @@ pub fn is_valid_indexed_attestation<
         ));
     }
 
-    let is_sorted = attesting_indices
-        .windows(2)
-        .map(|pair| {
-            let a = &pair[0];
-            let b = &pair[1];
-            a < b
-        })
-        .all(|x| x);
-    if !is_sorted {
-        return Err(invalid_operation_error(
-            InvalidOperation::IndexedAttestation(
-                InvalidIndexedAttestation::AttestingIndicesNotSorted,
-            ),
-        ));
-    }
-
-    let indices: HashSet<usize> = HashSet::from_iter(attesting_indices.iter().cloned());
-    if indices.len() != indexed_attestation.attesting_indices.len() {
-        let mut seen = HashSet::new();
-        let mut duplicates = vec![];
-        for i in indices.iter() {
-            if seen.contains(i) {
-                duplicates.push(*i);
-            } else {
-                seen.insert(i);
-            }
+    // List of indices is non-empty given check above. Begin iteration from one because a list with
+    // a single entry is "sorted" and contains no duplicates.
+    let mut prev = attesting_indices[0];
+    let mut duplicates = HashSet::new();
+    for index in attesting_indices.as_slice()[1..].iter() {
+        if *index < prev {
+            return Err(invalid_operation_error(
+                InvalidOperation::IndexedAttestation(
+                    InvalidIndexedAttestation::AttestingIndicesNotSorted,
+                ),
+            ));
+        }
+        if *index == prev {
+            duplicates.insert(*index);
         }
+        prev = *index;
+    }
+    if !duplicates.is_empty() {
         return Err(invalid_operation_error(
             InvalidOperation::IndexedAttestation(InvalidIndexedAttestation::DuplicateIndices(
-                duplicates,
+                Vec::from_iter(duplicates.into_iter()),
             )),
         ));
     }
+
     let mut public_keys = vec![];
-    for index in indices {
+    for index in attesting_indices.iter() {
```

Comment:
this should work the same...

```suggestion
    for index in attesting_indices {
```

Reply:
huh, without `.iter()` I get a compiler error that `ssz_rs::List` does not implement `Iterator`. I took a slice and iterate over that, but I'm not sure whether that's preferable to the original.

Reply:
hmm yeah, honestly I think `ssz_rs::List` should implement this but this is part of a wider push to `v1.0.0` for that library

here's a tracking issue: https://github.com/ralexstokes/ssz-rs/issues/52

let's either make a `TODO` comment here or hope that clippy will tell us it is unnecessary once we update the ssz lib

Reply:
having made another pass on this, I have decided to explicitly have callers use the methods (e.g. `.iter()`) they want in lieu of adding all of the iterator machinery for the possible things you may want to do